[AutoDiff] TF-309, TF-319: Handle `@noDerivative` properties.

dan-zheng opened the pull request:
- Change derived conformances to correctly mark non-differentiable stored
  properties with `@noDerivative`, not the parent nominal struct.
- Skip variedness propagation for `struct_element_addr` instructions to
  `@noDerivative` properties. Identical logic was previously added for
  `struct_extract`.
- Bail out in `AdjointEmitter::visitStructExtractInst` if the original
  field is `@noDerivative`.
- Add tests to ensure reproducer programs no longer crash.

Resolves [TF-309](https://bugs.swift.org/projects/TF/issues/TF-309) and [TF-319](https://bugs.swift.org/projects/TF/issues/TF-319).

dan-zheng committed:
[AutoDiff] TF-309, TF-319: Handle `@noDerivative` properties.

- Change derived conformances to correctly mark non-differentiable stored
  properties with `@noDerivative`, not the parent nominal struct.
- Skip variedness propagation for `struct_element_addr` instructions to
  `@noDerivative` properties. Identical logic was previously added for
  `struct_extract`.
- Bail out in `AdjointEmitter::visitStructExtractInst` if the original
  field is `@noDerivative`.
- Add tests to ensure reproducer programs no longer crash.

Resolves TF-309 and TF-319.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1346,6 +1346,21 @@ void DifferentiableActivityInfo::analyze(DominanceInfo *di,
           }
         }
+        // Handle `struct_element_addr`.
+        else if (auto *seai = dyn_cast<StructElementAddrInst>(&inst)) {
```
The body is exactly the same as SEI's except for `seai`/`sei`. Could you write a macro?

dan-zheng replied:
Macro'd in https://github.com/apple/swift/pull/22973/commits/272e986ca9dbf0faaa6040e24b0d34d57c9188c5: `#define DIAGNOSE_STRUCT_NO_DERIVATIVE_FIELD(INST)`.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4229,6 +4244,10 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   void visitStructExtractInst(StructExtractInst *sei) {
+    // There does not exist a corresponding cotangent field for original
+    // fields with `@noDerivative` attribute. Emit an error.
+    if (sei->getField()->getAttrs().hasAttribute<NoDerivativeAttr>())
```
I don't think you need this anymore. I believe an assertion is safe.

dan-zheng replied:
Assertion added in https://github.com/apple/swift/pull/22973/commits/37c6620662c267ae2cd8665f4d8d620fbec8192a.

Verified with sample programs that assertion is never triggered, because activity analysis does not propagate variedness for `struct_extract` instructions with `@noDerivative` fields.

dan-zheng committed:
Simplify adjoint projection logic.

- Diagnose `@noDerivative` stored properties instead of
  non-differentiable buffers in `getAdjointBuffer`. Add test.
- Move projection-handling code from `getAdjointBuffer` into a separate
  function: `getAdjointProjection`.
- Create `DIAGNOSE_STRUCT_NO_DERIVATIVE_FIELD` macro to reduce code
  dupe for `StructExtractInst` and `StructElementAddrInst`.
- Create `PROPAGATE_CLEANUP` macro to reduce code dupe for
  `StructElementAddrInst` and `TupleElementAddrInst`.

dan-zheng committed:
Emit warning+fixit for non-varied results, add assertion.

- Emit warning+fixit when original result is not varied because the
  derivative is always zero. Add tests.
- Assert in `AdjointEmitter::visitStructExtractInst` that struct field
  is not `@noDerivative`.

dan-zheng committed:
Fix macro doc comments.

dan-zheng commented:
Ready for re-review. An additional diagnostic and tests have been added.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1331,21 +1331,29 @@ void DifferentiableActivityInfo::analyze(DominanceInfo *di,
+          setVaried(result, i); \
+    } \
+  }
+
```
Remove empty line.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22973/commits/e42706e5b266d61589a06dcb3ee67cfc1a2ac407.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1331,21 +1331,29 @@ void DifferentiableActivityInfo::analyze(DominanceInfo *di,
+//   `@_fieldwiseDifferentiable` struct, do not set the result as varied because
+//   it is not in the set of differentiable variables.
+// - Otherwise, propagate variedness from operand to result as usual.
+#define VISIT_STRUCT_ELEMENT_INST(INST) \
```
```suggestion
#define PROPAGATE_VARIED_FOR_STRUCT_EXTRACTION(INST) \
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22973/commits/e42706e5b266d61589a06dcb3ee67cfc1a2ac407.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3803,6 +3807,17 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+    // a zero derivative.
+    if (!activityInfo.isVaried(origResult, task->getIndices().source)) {
+      // Emit fixit if original result has a valid source location.
+      auto sourceLoc = origResult.getLoc().getSourceLoc();
```
Does it have an `operator bool()`? Can you do `if (auto sourceLoc = ...` instead?

dan-zheng replied:
It seems not.
`Value of type 'swift::SourceLoc' is not contextually convertible to 'bool'`

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4562,6 +4580,17 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
                      ValueWithCleanup(adjAccess, makeCleanupFromChildren({})));
   }
 
+#define PROPAGATE_CLEANUP(INST) \
```
How about
```suggestion
#define PROPAGATE_BUFFER_CLEANUP(INST) \
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22973/commits/e42706e5b266d61589a06dcb3ee67cfc1a2ac407.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4562,6 +4580,17 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+#define PROPAGATE_CLEANUP(INST) \
+  void visit##INST##Inst(INST##Inst *inst) { \
+    auto adjBase = getAdjointBuffer(inst->getOperand()); \
+    auto adjProj = getAdjointBuffer(inst); \
```
If it's not a reference then `setCleanup` has no effect.

```suggestion
    auto &adjProj = getAdjointBuffer(inst); \
```

dan-zheng replied:
Thanks for the catch. Done in https://github.com/apple/swift/pull/22973/commits/e42706e5b266d61589a06dcb3ee67cfc1a2ac407.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4562,6 +4580,17 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
 
+#define PROPAGATE_CLEANUP(INST) \
+  void visit##INST##Inst(INST##Inst *inst) { \
+    auto adjBase = getAdjointBuffer(inst->getOperand()); \
```
reference as well

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22973/commits/e42706e5b266d61589a06dcb3ee67cfc1a2ac407.

rxwei commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -375,6 +375,12 @@ NOTE(autodiff_nondifferentiable_argument,none,
+     "cannot differentiate through a '@noDerivative' stored property; do you "
+     "want to add '.withoutDerivative()'?", ())
+WARNING(autodiff_nonvaried_result_fixit,none,
+        "result does not depend on differentiation arguments and always has a "
```
```suggestion
        "result does not depend on differentiation arguments and will always have a "
```

dan-zheng replied:
Updated in https://github.com/apple/swift/pull/22973/commits/e42706e5b266d61589a06dcb3ee67cfc1a2ac407.

rxwei reviewed (approved):
LGTM!

Future cleanup suggestion: Make `getAdjointBuffer` return an `Optional` so that the caller can check that instead of checking `errorOccurred`. Maybe rename it to `tryRetrieveAdjointBuffer()` or something.

dan-zheng committed:
Minor edits.

Address feedback from @rxwei.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
